Modernize PInvoke with CSWin32

AliveDevil opened the pull request:
First part of #12805.

Fixes #12803 
Fixes #12804

AliveDevil committed:
Add CsWin32

AliveDevil committed:
Remove AnyCPU configuration

AliveDevil committed:
Update Credentials to CsWin32

AliveDevil committed:
Update manual PInvoke to upstream CsWin32

AliveDevil committed:
Prepare TaskDialog

AliveDevil committed:
Prepare TaskDialog

AliveDevil committed:
Update TaskDialog

AliveDevil committed:
Redundant

AliveDevil committed:
Prepare TaskDialog

AliveDevil committed:
Refresh

AliveDevil committed:
TaskDialog

AliveDevil committed:
Use CSWin32 daily

AliveDevil committed:
use CsWin32 for Windows

AliveDevil committed:
Update PInvoke

AliveDevil committed:
Remove unused

AliveDevil committed:
Port VirtualDesktop support

AliveDevil committed:
Port IInitializeWithWindow

AliveDevil committed:
Update

AliveDevil committed:
Review DragDrop

AliveDevil committed:
Public available nuget version

AliveDevil committed:
Fix TaskDialog

AliveDevil committed:
Fix certificate window

AliveDevil committed:
Drop CredentialManagerDefaultFallback

AliveDevil committed:
Review CredentialManagerPasswordStore

AliveDevil committed:
Dont explicitely implement interface. Override instead

AliveDevil commented on core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs:
```diff
@@ -49,30 +48,178 @@ public override void addPassword(Scheme scheme, int port, string hostName, strin
+            {
+                winCred.Attributes["Private Key Passphrase"] = credential.getIdentityPassphrase();
+            }
+            if (credential.isPasswordAuthentication())
```
Need to check here.
If password authentication is required, then if either username or password is empty the whole credential is not saved.

Can private key configurable be used with password authentication, and username is empty?

dkocher replied:
Should match logic in `DefaultHostPasswordStore` to make sure we have no regressions.

AliveDevil replied:
DefaultHostPasswordStore runs separate addPassword-calls for these conditions.

CredentialManagerPasswordStore now only performs one single commit with all fields combined.

AliveDevil commented on core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs:
```diff
@@ -49,30 +48,178 @@ public override void addPassword(Scheme scheme, int port, string hostName, strin
+                }
+            }
+            targetBuilder.Path = pathBuilder.ToString();
+            targetBuilder.Query = "user=" + credentials.getUsername();
```
Uses format `application.container.name:protocol.identifier(:port)?\?user=username`.
Should this be changed/updated?

AliveDevil replied:
Upping this one.

`application.container.name` for Cyberduck is `duck`, for Mountain Duck Windows this is `Cyberduck`. This way Cyberduck and Mountain Duck don't share credentials, which is kind of unfortunate. Which key should be used for these shared elements? Taking into account branded builds.

AliveDevil commented:
Tests won't compile, need to fix that.

AliveDevil committed:
Fix tests

AliveDevil committed:
Fix tests

AliveDevil committed:
Add huge blob test

dkocher commented on core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs:
```diff
@@ -40,7 +39,7 @@ public override void addPassword(string serviceName, string user, string passwor
             if (!WinCredentialManager.SaveCredentials(hostUrl, new NetworkCredential(user, password)))
             {
-                throw new LocalAccessDeniedException($"Could not save credentials for \"{hostUrl}\" to Windows Credential Manager.");
+                logger.error($"Could not save credentials for \"{hostUrl}\" to Windows Credential Manager.");
```
Why no longer throwing failure?

AliveDevil replied:
Is there a difference in throwing? Will be written into error log regardless, and prevents unecessary stack rewinds and exception handling.

dkocher commented on core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs:
```diff
@@ -17,15 +17,14 @@
+using static Windows.Win32.Security.Credentials.CRED_TYPE;
 
 namespace Ch.Cyberduck.Core
 {
```
Should probably just implement `HostPasswordStore` if overriding all implementations from `DefaultHostPasswordStore` anyways.

AliveDevil replied:
I'd agree, unless there is logic in DefaultHostPasswordStore, that I don't want to replicate _again_. Thats why I overwrote the HostPasswordStore-implementation for the revised implementation, and falling back to DefaultHostPasswordStore-logic.
See original logic for CredentialManagerPasswordStore: https://github.com/iterate-ch/cyberduck/blob/master/core/src/main/csharp/ch/cyberduck/core/CredentialManagerPasswordStore.cs
I didn't want to have three different proxies in https://github.com/iterate-ch/cyberduck/blob/feature/GH-12805-1/core/src/main/csharp/ch/cyberduck/core/PasswordStoreFacade.cs.

AliveDevil committed:
Review

AliveDevil committed:
Don't fail when username or password is empty - if username is empty WinCredentialManager.SaveCredentials() fails, when Password is empty just print a warning.

AliveDevil committed:
Review

AliveDevil committed:
Fix Build

AliveDevil committed:
Always build Test parallel to Windows project to verify tests would atleast build successfully.

AliveDevil committed:
Fix build

AliveDevil committed:
Update .editorconfig

AliveDevil committed:
Reformat project files

AliveDevil committed:
Make builds more resilient

AliveDevil committed:
Remove _all_ x86 (legacy WinForms Designer support) configurations.
"Windows" configuration builds everything

AliveDevil committed:
Installer

AliveDevil committed:
Update dependencies

AliveDevil committed:
Update CsWin32

AliveDevil committed:
Fix packagereference for CsWin32

AliveDevil committed:
Fix build

AliveDevil commented:
Works For Me (tm)
```
[INFO]   [msbuild] C:\Program Files\Microsoft Visual Studio\2022\Enterprise\MSBuild\Microsoft\VC\v170\Microsoft.CppBuild.targets(437,5): error MSB8013: This project doesn't contain the Configuration and Platform combination of Release|x64. [D:\a\cyberduck\cyberduck\windows\src\main\wix\WindowsVersionCustomAction\WindowsVersionCustomAction.vcxproj]
```

tatze96 commented:
[637825876270682181.txt](https://github.com/iterate-ch/cyberduck/files/8230295/637825876270682181.txt)
Cyberduck crasht beim verwenden des Disconnect Buttons.

AliveDevil committed:
Enforce Platform in Traversal project

AliveDevil committed:
Fix null references

AliveDevil committed:
Review retrieval/saving of credentials
Using Unicode for everything (reduces possible length of passwords to 256 characters)

AliveDevil committed:
Mismatch in IDs

AliveDevil committed:
Fix DragDrop

AliveDevil committed:
Review Task Dialog. Can use up to 128 command buttons

AliveDevil committed:
Move CsWin32-generated interface to custom written

AliveDevil committed:
Fix Drag/Drop

AliveDevil committed:
Review dialogs

AliveDevil commented:
Wouldn't merge for 8.3.

AliveDevil committed:
Update implementation for SHGetFileInfo

AliveDevil committed:
Add ref char constructor to PWSTR

AliveDevil committed:
Limit tostring

AliveDevil committed:
Update Credentials to CsWin32

AliveDevil committed:
Fix UI

AliveDevil committed:
Auto trim in View

AliveDevil committed:
Make use of new duck-URI format in Credential Manager for BookmarkController
